packages/edit-widgets/src/components/widget-area/selection-observer.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withSelect, withDispatch } from '@wordpress/data';
+
+/**
+ * Component which calls onBlockSelected prop when a block becomes selected. It
+ * can be used to ensuring that only one block appears as selected
+ * when multiple editors exist in a page.
+ *
+ * @type {WPComponent}
+ */
+
+class SelectionObserver extends Component {
+	componentDidUpdate( prevProps ) {
+		const {
+			hasSelectedBlock,
+			onBlockSelected,
+			isSelectedArea,
+			clearSelectedBlock,
+		} = this.props;
+
+		if ( hasSelectedBlock && ! prevProps.hasSelectedBlock ) {
+			onBlockSelected();
+		}
+
+		if ( ! isSelectedArea && prevProps.isSelectedArea ) {
+			clearSelectedBlock();
+		}
+	}
+
+	render() {
+		return null;
+	}
+}
+
+export default compose( [
+	withSelect( ( select ) => {
```

Comment:
useSelect, useDispatch? ;)

Reply:
I tried to use hooks for this component but hooks did not work as expected, I always got unexpected results.
I tried to use a simple version where I relied on effects that execute when prop changes, and I also tried a usePrevious hook https://usehooks.com/usePrevious/ that allowed an implementation almost equal to this class component version, and both versions did not work properly. It is a very strange case I may be missing something.